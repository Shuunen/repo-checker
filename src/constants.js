import { join } from 'path'

export const home = process.env.HOME
export const dataFileName = 'repo-checker.config.js'
export const repoCheckerPath = join(__dirname, '..')
export const templatePath = join(repoCheckerPath, 'templates')
export const dataFileHomePath = join(home, dataFileName)

export const dataDefaults = {
  auto_merge: true,
  ban_sass: true,
  dev_deps_only: true,
  license: 'GPL-3.0',
  max_size_ko: 10,
  user_mail: 'romain.racamier@gmail.com',
  user_name: 'Romain Racamier-Lafon',
}
