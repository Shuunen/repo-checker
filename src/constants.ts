import { join } from 'path'

/* istanbul ignore next */
export const home = process.env.HOME ?? '' // when does HOME is not defined ?
export const dataFileName = 'repo-checker.config.js'
/* istanbul ignore next */
export const repoCheckerPath = process.env.pwd ?? process.cwd()
export const templatePath = join(repoCheckerPath, 'templates')
export const dataFileHomePath = join(home, dataFileName)

export class ProjectData {
  auto_merge = true
  ban_sass = true
  dev_deps_only = true
  is_module = false
  license = 'GPL-3.0'
  max_size_ko = 1
  npm_package = false
  package_name = ''
  quiet = false
  repo_id = 'a-great-repo'
  use_typescript = false
  use_vue = false
  user_id = 'Shuunen'
  user_id_lowercase = 'shuunen'
  user_mail = 'romain.racamier@gmail.com'
  user_name = 'Romain Racamier-Lafon'
  web_published = false
  web_url = 'https://my-website.com'

  constructor (data: Partial<ProjectData> = {}) {
    Object.assign(this, data)
  }
}

export const dataDefaults = new ProjectData()
