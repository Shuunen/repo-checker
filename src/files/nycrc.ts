import { FileBase } from '../file.ts'
import { log } from '../logger.ts'

// eslint-disable-next-line no-restricted-syntax, jsdoc/require-jsdoc
export class NycRcFile extends FileBase {
  // eslint-disable-next-line jsdoc/require-jsdoc
  private async getConfigFileName() {
    const hasRc = await this.fileExists('.nycrc')
    const hasRcJson = await this.fileExists('.nycrc.json')
    const hasConfigFile = hasRc || hasRcJson
    this.test(hasConfigFile, 'nycrc file exists')
    if (!hasConfigFile) return ''
    /* c8 ignore next */
    return hasRc ? '.nycrc' : '.nycrc.json'
  }

  // eslint-disable-next-line jsdoc/require-jsdoc
  public async start() {
    if (!this.data.isUsingNyc) {
      log.debug('does not use nyc')
      return
    }
    const fileName = await this.getConfigFileName()
    if (fileName === '') return
    await this.checkFileExists(fileName)
    await this.inspectFile(fileName)
    this.couldContainsSchema('https://json.schemastore.org/nycrc')
  }
}
