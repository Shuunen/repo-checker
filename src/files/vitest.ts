import { FileBase } from '../file.ts'

// eslint-disable-next-line no-export
export class VitestFile extends FileBase {
  // eslint-disable-next-line jsdoc/require-jsdoc
  // eslint-disable-next-line max-lines-per-function
  public async start() {
    if (!this.data.isUsingVitest) return
    const hasMtsFile = await this.fileExists('vitest.config.mts')
    const hasJsFile = await this.fileExists('vitest.config.js')
    const hasTsFile = await this.fileExists('vitest.config.ts')
    const hasMjsFile = await this.fileExists('vitest.config.mjs')
    if (hasMtsFile) await this.inspectFile('vitest.config.mts')
    /* c8 ignore next 3 */
    if (hasJsFile) await this.inspectFile('vitest.config.js')
    if (hasTsFile) await this.inspectFile('vitest.config.ts')
    if (hasMjsFile) await this.inspectFile('vitest.config.mjs')
    this.couldContains('a pool thread setting', /pool: 'threads'/u)
  }
}
