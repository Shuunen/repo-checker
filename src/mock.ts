/* eslint-disable jsdoc/require-jsdoc */
import path from 'node:path'
import { clone, sleep } from 'shuutils'
import type { Indicators } from './check.ts'
import type { FileBase } from './file.ts'
import { join } from './utils.ts'

const fileBaseKeysToDelete: (keyof FileBase)[] = ['fileExists', 'checkFileExists', 'folderPath', 'inspectFile', 'updateFile', 'initFile']
const indicatorsKeys: ReadonlyArray<keyof Indicators> = ['failed', 'passed', 'warnings'] as const

function cleanStringForSnap(input: string) {
  return (
    input
      // replace repo-check-1-40 by repo-check-x-yy
      .replace(/repo-check-\d-\d+/gu, 'repo-check-x-yy')
      // replace package-json-main-file-size-63ko by package-json-main-file-size-XYZko
      .replace(/package-json-main-file-size-\d+ko/gu, 'package-json-main-file-size-XYZko')
  )
}

// eslint-disable-next-line no-restricted-syntax
export function cleanUnknownValueForSnap<Type>(input: Type): Type {
  /* c8 ignore next 8 */
  /* eslint-disable @typescript-eslint/consistent-type-assertions, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-type-assertion */
  if (typeof input === 'string') return cleanStringForSnap(input) as Type
  if (Array.isArray(input)) return input.map(item => cleanUnknownValueForSnap(item)) as Type
  if (typeof input === 'object') return JSON.parse(cleanStringForSnap(JSON.stringify(input))) as Type
  // eslint-disable-next-line no-restricted-syntax
  throw new Error(`cleanInstanceValueForSnap: unknown type ${typeof input}`)
  /* eslint-enable @typescript-eslint/consistent-type-assertions, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-type-assertion */
}

export async function promiseValue<Type>(value: Type) {
  await sleep(1)
  return value
}

export async function promiseTrue() {
  await sleep(1)
  return true
}

export async function promiseFalse() {
  await sleep(1)
  return false
}

export async function promiseVoid() {
  await sleep(1)
}

// eslint-disable-next-line unicorn/prefer-module
export const sourceFolder = path.dirname(__filename)

export const vueProjectFolder = join(sourceFolder, 'mocks', 'vueProject')

export const tsProjectFolder = join(sourceFolder, 'mocks', 'tsProject')

export const mocksProjectsFolder = join(sourceFolder, 'mocks')

export function cleanIndicatorsForSnap(record?: Readonly<Record<string, unknown>>) {
  if (!record) return record
  const clean = clone(record)
  for (const key of indicatorsKeys) if (key in clean) clean[key] = cleanUnknownValueForSnap(clean[key])
  return clean
}

// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
export function cleanInstanceForSnap(instance: Readonly<FileBase>, ...bonusKeys: readonly string[]) {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  const clean = clone<Partial<FileBase>>(instance) as Record<string, unknown>
  const keys = [...fileBaseKeysToDelete, ...bonusKeys] // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
  for (const key of keys) if (key in clean) delete clean[key]
  return cleanIndicatorsForSnap(clean)
}
