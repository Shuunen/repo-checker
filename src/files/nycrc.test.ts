import { expect, it } from 'vitest'
import { ProjectData, repoCheckerPath } from '../constants'
import { cleanInstanceForSnap, promiseFalse, vueProjectFolder } from '../mock'
import { NycRcFile } from './nycrc'

it('nyc rc A file exists', async () => {
  const instance = new NycRcFile(repoCheckerPath, new ProjectData({ isQuiet: true, isUsingNyc: true }))
  await instance.start()
  await instance.end()
  expect(cleanInstanceForSnap(instance)).toMatchSnapshot()
})

it('nyc rc B check skip on a non nyc/c8 project', async () => {
  const instance = new NycRcFile(repoCheckerPath, new ProjectData({ isQuiet: true }))
  await instance.start()
  await instance.end()
  expect(cleanInstanceForSnap(instance)).toMatchSnapshot()
})

it('nyc rc C file missing', async () => {
  const instance = new NycRcFile(repoCheckerPath, new ProjectData({ isQuiet: true, isUsingNyc: true }))
  instance.fileExists = promiseFalse
  await instance.start()
  await instance.end()
  expect(cleanInstanceForSnap(instance)).toMatchSnapshot()
})

it('nyc rc D file exists but not a json ext file', async () => {
  const instance = new NycRcFile(vueProjectFolder, new ProjectData({ isQuiet: true, isUsingNyc: true }))
  await instance.start()
  await instance.end()
  expect(cleanInstanceForSnap(instance)).toMatchSnapshot()
})
