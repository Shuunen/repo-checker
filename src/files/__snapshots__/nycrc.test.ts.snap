// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`nyc rc A file exists 1`] = `
{
  "canFix": false,
  "canForce": false,
  "data": {
    "canAutoMergeDeps": true,
    "hasMainBranch": false,
    "hasTaskPrefix": false,
    "isCli": false,
    "isModule": false,
    "isPublishedPackage": false,
    "isQuiet": true,
    "isUsingC8": false,
    "isUsingDependencyCruiser": false,
    "isUsingEslint": false,
    "isUsingNyc": true,
    "isUsingShuutils": false,
    "isUsingTailwind": false,
    "isUsingTypescript": false,
    "isUsingV8": false,
    "isUsingVue": false,
    "isWebPublished": false,
    "license": "GPL-3.0",
    "maxSizeKo": 1,
    "packageName": "",
    "repoId": "a-great-repo",
    "shouldAvoidSass": true,
    "userId": "Shuunen",
    "userIdLowercase": "shuunen",
    "userMail": "romain.racamier@gmail.com",
    "userName": "Romain Racamier-Lafon",
    "webUrl": "https://my-website.com",
    "willGenerateReport": true,
  },
  "failed": [
    "nycrc-file-exists",
  ],
  "fileContent": "",
  "fileName": "",
  "originalFileContent": "",
  "passed": [],
  "warnings": [],
}
`;

exports[`nyc rc B check skip on a non nyc/c8 project 1`] = `
{
  "canFix": false,
  "canForce": false,
  "data": {
    "canAutoMergeDeps": true,
    "hasMainBranch": false,
    "hasTaskPrefix": false,
    "isCli": false,
    "isModule": false,
    "isPublishedPackage": false,
    "isQuiet": true,
    "isUsingC8": false,
    "isUsingDependencyCruiser": false,
    "isUsingEslint": false,
    "isUsingNyc": false,
    "isUsingShuutils": false,
    "isUsingTailwind": false,
    "isUsingTypescript": false,
    "isUsingV8": false,
    "isUsingVue": false,
    "isWebPublished": false,
    "license": "GPL-3.0",
    "maxSizeKo": 1,
    "packageName": "",
    "repoId": "a-great-repo",
    "shouldAvoidSass": true,
    "userId": "Shuunen",
    "userIdLowercase": "shuunen",
    "userMail": "romain.racamier@gmail.com",
    "userName": "Romain Racamier-Lafon",
    "webUrl": "https://my-website.com",
    "willGenerateReport": true,
  },
  "failed": [],
  "fileContent": "",
  "fileName": "",
  "originalFileContent": "",
  "passed": [],
  "warnings": [],
}
`;

exports[`nyc rc C file missing 1`] = `
{
  "canFix": false,
  "canForce": false,
  "data": {
    "canAutoMergeDeps": true,
    "hasMainBranch": false,
    "hasTaskPrefix": false,
    "isCli": false,
    "isModule": false,
    "isPublishedPackage": false,
    "isQuiet": true,
    "isUsingC8": false,
    "isUsingDependencyCruiser": false,
    "isUsingEslint": false,
    "isUsingNyc": true,
    "isUsingShuutils": false,
    "isUsingTailwind": false,
    "isUsingTypescript": false,
    "isUsingV8": false,
    "isUsingVue": false,
    "isWebPublished": false,
    "license": "GPL-3.0",
    "maxSizeKo": 1,
    "packageName": "",
    "repoId": "a-great-repo",
    "shouldAvoidSass": true,
    "userId": "Shuunen",
    "userIdLowercase": "shuunen",
    "userMail": "romain.racamier@gmail.com",
    "userName": "Romain Racamier-Lafon",
    "webUrl": "https://my-website.com",
    "willGenerateReport": true,
  },
  "failed": [
    "nycrc-file-exists",
  ],
  "fileContent": "",
  "fileName": "",
  "originalFileContent": "",
  "passed": [],
  "warnings": [],
}
`;

exports[`nyc rc D file exists but not a json ext file 1`] = `
{
  "canFix": false,
  "canForce": false,
  "data": {
    "canAutoMergeDeps": true,
    "hasMainBranch": false,
    "hasTaskPrefix": false,
    "isCli": false,
    "isModule": false,
    "isPublishedPackage": false,
    "isQuiet": true,
    "isUsingC8": false,
    "isUsingDependencyCruiser": false,
    "isUsingEslint": false,
    "isUsingNyc": true,
    "isUsingShuutils": false,
    "isUsingTailwind": false,
    "isUsingTypescript": false,
    "isUsingV8": false,
    "isUsingVue": false,
    "isWebPublished": false,
    "license": "GPL-3.0",
    "maxSizeKo": 1,
    "packageName": "",
    "repoId": "a-great-repo",
    "shouldAvoidSass": true,
    "userId": "Shuunen",
    "userIdLowercase": "shuunen",
    "userMail": "romain.racamier@gmail.com",
    "userName": "Romain Racamier-Lafon",
    "webUrl": "https://my-website.com",
    "willGenerateReport": true,
  },
  "failed": [],
  "fileContent": "{
  "$schema": "https://json.schemastore.org/nycrc",
  "branches": 90,
  "check-coverage": true,
  "functions": 90,
  "include": ["src/**/*.ts"],
  "lines": 90,
  "reporter": ["html", "text", "lcovonly"],
  "statements": 90
}
",
  "fileName": ".nycrc",
  "originalFileContent": "{
  "$schema": "https://json.schemastore.org/nycrc",
  "branches": 90,
  "check-coverage": true,
  "functions": 90,
  "include": ["src/**/*.ts"],
  "lines": 90,
  "reporter": ["html", "text", "lcovonly"],
  "statements": 90
}
",
  "passed": [
    "nycrc-file-exists",
    "has-a-nycrc-file",
    "nycrc-has-a-schema-declaration",
  ],
  "warnings": [],
}
`;
